### PR TITLE
filesystem api

### LINKED ISSUE #2119: Create API endpoints to store arbitrary data
The Angular wallet sometimes needs to save data. However, it is not possible to use `LocalStorage`, because the backend port changes in each execution, making the data inaccessible. Also, it would be good not to have to use Electron to access the file system.

To solve this it is necessary to create API endpoints to save data and read it. At least at the moment something very complex is not required, just being able to store arbitrary data (preferably a text string, to be able to write and read it easily from Javascript) and be able to identify the data in some way (probably an string serving as ID).

@mahansky , do you have any special requirements regarding this?

### PR DESCRIPTION
**Pull request by therealssj**
Fixes #2119 

Added `api/v2/data` endpoint which lets you store, get, delete or update data in a file. The url uses HTTP verbs (POST, GET and DELETE) for different actions.

Changes:
- Add `-data-file-path` option to set path of the arbitrary data store file. Defaults to `~/.skycoin/data.json`.
- Add `/api/v2/data` endpoint to store, get, delete or update arbitrary data in a file on disk defined by the `-data-file-path` option.
- Add new api set `FILESYSTEM` to enable `/api/v2/data` endpoint.

Does this change need to mentioned in CHANGELOG.md?
Yes

### CODE REVIEW

**Comment by gz-c** [comment]
Use HTTP verbs

**Comment by gz-c** [comment]
Use HTTP verbs:

File: Gopkg.lock [hunk 1]
```diff
@@ -145,17 +145,17 @@
   branch = "master"
   name = "github.com/prometheus/client_model"
   packages = ["go"]
-  revision = "5c3871d89910bfb32f5fcab2aa4b9ec68e65a99f"
+  revision = "fd36f4220a901265f90734c3183c5f0c91daa0b8"
```
Comment (gz-c) [inline Gopkg.lock]: Was a new dep added? I don't see one; if not, don't update the deps yet, unless its necessary, but you can do it in another PR

File: src/api/README.md [hunk 2]
```diff
@@ -1539,6 +1543,126 @@ Result:
...
 ```
 
+## Filesystem APIs
+
+### Save data
+
+API sets: `FILESYSTEM`
+
+```
+URI: /api/v2/data/save
```
Comment (gz-c) [inline src/api/README.md]: The endpoints still aren't updated to use HTTP verbs

File: src/api/filesystem.go [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+	Filename string                 `json:"filename"`
+	Data     map[string]interface{} `json:"data"`
+	Update   bool                   `json:"update"`
+}
+
+// Save arbitrary data to disk
+// URI: /api/v2/data
+// Method: POST
+// Args:
+//     filename: filename [required]
```
Comment (gz-c) [inline src/api/filesystem.go]: Filename should not be part of the API, use a single known filename, like data.db uses. Consider this another database file

**Comment by gz-c** [comment]
Can you update the PR description to reflect the URLs

File: src/api/README.md [hunk 4]
```diff
@@ -1648,6 +1652,119 @@ Result:
...
+Args:
+    keys: list of keys to retrieve [required]
+```
+
+Get data from data file on disk
+
+Example:
+
+```sh
+curl -X GET \
```
Comment (gz-c) [inline src/api/README.md]: Isn't `-F` for formdata? Should be a query string

File: src/api/README.md [hunk 5]
```diff
@@ -1648,6 +1652,119 @@ Result:
...
+```
+
+Delete data from data file on disk
+
+Example:
+
+```sh
+curl -X DELETE \
+  http://127.0.0.1:6420/api/v2/data \
+  -F 'keys=key1,key2,key3'
```
Comment (gz-c) [inline src/api/README.md]: Should be query string

File: src/api/filesystem.go [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+
+// Save arbitrary data to disk
+// URI: /api/v2/data
+// Method: POST
+// Args:
+//     data: arbitrary data to save [required]
+//     update: update existing values [optional]
+func dataSaveHandler(gateway Gatewayer, w http.ResponseWriter, r *http.Request) {
+	if r.Method != http.MethodPost {
+		wh.Error405(w)
```
Comment (gz-c) [inline src/api/filesystem.go]: wrong type of error responses for v2 API

File: src/api/filesystem.go [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+	}
+
+	err = gateway.SaveData(params.Data, params.Update)
+	if err != nil {
+		resp := NewHTTPErrorResponse(http.StatusInternalServerError, err.Error())
+		writeHTTPResponse(w, resp)
+		return
+	}
+
+	wh.SendJSONOr500(logger, w, "success")
```
Comment (gz-c) [inline src/api/filesystem.go]: Wrong type of response for v2 api.
No need to include "success" message; leave it empty. The `200 OK` is sufficient to indicate success

File: src/api/README.md (hunk 4 above)
Comment (therealssj) [reply src/api/README.md]: both work apparently. I been using form data only in postman

File: src/daemon/gateway.go [hunk 8]
```diff
@@ -541,6 +543,69 @@ func (gw *Gateway) GetWalletSeed(id string, password []byte) (string, error) {
 	return gw.v.Wallets.GetWalletSeed(id, password)
 }
 
+// SaveData saves arbitrary data to data file on disk
+func (gw *Gateway) SaveData(data map[string]interface{}, update bool) error {
```
Comment (gz-c) [inline src/daemon/gateway.go]: Don't put logic in the gateway. Gateway is only a shim to give the api package an interface for mocking

File: src/skycoin/config.go [hunk 9]
```diff
@@ -188,6 +188,9 @@ type NodeConfig struct {
 	LogToFile   bool
 	Version     bool // show node version
 
+	// DataFilePath path of the file to store arbitrary data on disk
+	DataFilePath string
```
Comment (gz-c) [inline src/skycoin/config.go]: `Data` is ambiguous, we need better terminology. Maybe `Keystore`

File: src/skycoin/config.go [hunk 10]
```diff
@@ -535,7 +544,7 @@ func buildAPISets(c NodeConfig) (map[string]struct{}, error) {
 		api.EndpointsTransaction,
 		api.EndpointsPrometheus,
 		api.EndpointsNetCtrl,
-		// Do not include insecure or deprecated API sets, they must always
+		// Do not include insecure, deprecated or filesystem API sets, they must always
```
Comment (gz-c) [inline src/skycoin/config.go]: I think the filesystem API can be left in the main grouping

File: src/api/filesystem.go [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+	var params saveDataRequest
+	err := decoder.Decode(&params)
+	if err != nil {
+		resp := NewHTTPErrorResponse(http.StatusBadRequest, err.Error())
+		writeHTTPResponse(w, resp)
+		return
+	}
+	defer r.Body.Close()
+
+	if params.Data == nil {
```
Comment (gz-c) [inline src/api/filesystem.go]: len(params.Data) == 0

File: src/daemon/gateway.go (hunk 8 above)
Comment (gz-c) [reply src/daemon/gateway.go]: Belongs in its own package. Furthermore, none of this code is tested?

Comment (gz-c) [reply src/daemon/gateway.go]: The `update` flag makes no sense. If `update` is false it is a no-op?

Comment (gz-c) [reply src/daemon/gateway.go]: I removed daemon.Gateway; the new way to do this is:

- New package for managing the keystore
- New instance of keystore manager created during coin init
- Instance of keystore manager passed to api.NewGateway
- `Keystorer` interface added to api/gateway.go

**Comment by gz-c** [comment]
Will be handled byb #2223
